sdk/containerregistry/azure-containerregistry/azure/containerregistry/_download_stream.py

```diff
@@ -0,0 +1,70 @@
+# coding=utf-8
+# ------------------------------------
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT License.
+# ------------------------------------
+import hashlib
+from typing import Iterator, ContextManager, Callable, TypeVar, cast, Tuple, Dict
+
+T = TypeVar('T', bound='DownloadBlobStream')
+
+
+class DownloadBlobStream(
+    Iterator[bytes],
+    ContextManager[T],
+):
+    """Protocol for methods to provide streamed responses."""
+
+    def __init__(
+        self,
+        *,
+        response: Iterator[bytes],
+        get_next: Callable[[str], Iterator[bytes]],
+        blob_size: int,
+        downloaded: int,
+        digest: str,
+        chunk_size: int
+    ) -> None:
+        self._response = response
+        self._next = get_next
+        self._blob_size = blob_size
+        self._downloaded = downloaded
+        self._digest = digest
+        self._chunk_size = chunk_size
+        self._hasher = hashlib.sha256()
+
+    def __enter__(self: T) -> T:
+        return self
+
+    def __exit__(self, *args) -> None:
+        return None
+
+    def __iter__(self: T) -> T:
+        return self
+
+    def _yield_data(self) -> bytes:
+        data = next(self._response)
+        self._hasher.update(data)
+        return data
+
+    def _download_chunk(self) -> Iterator[bytes]:
+        end_range = self._downloaded + self._chunk_size
+        range_header = f"bytes={self._downloaded}-{end_range}"
+        next_chunk, headers = cast(
+            Tuple[Iterator[bytes], Dict[str, str]],
+            self._next(range=range_header) # type: ignore
```

Comment:
Is it that mypy yells about the keyword? typing doesn't have a great way to handle callables with keyword only arguments. You could try a [callback protocol](https://peps.python.org/pep-0544/#callback-protocols), something like:

```python
from typing_extensions import Protocol

class GetNext(Protocol):
    def __call__(self, *values: str, range: str) -> Iterator[bytes]: ...
```